synapse/rest/admin/registration_tokens.py

```diff
@@ -0,0 +1,195 @@
+            if not set(token).issubset(self.allowed_chars_set):
+                raise SynapseError(
+                    400,
+                    "token must consist only of characters matched by the regex [A-Za-z0-9-_]",
+                    Codes.INVALID_PARAM,
+                )
+
+        else:
+            # Get length of token to generate (default is 16)
+            length = body.get("length", 16)
+            if not isinstance(length, int):
+                raise SynapseError(
+                    400, "length must be an integer", Codes.INVALID_PARAM
+                )
+            if length > 64:
+                raise SynapseError(
+                    400, "length must not be greater than 64", Codes.INVALID_PARAM
+                )
+
+            # Generate token
+            token = "".join(random.choices(self.allowed_chars, k=length))
+
+        uses_allowed = body.get("uses_allowed", None)
+        if not isinstance(uses_allowed, (int, type(None))):
+            raise SynapseError(
+                400, "uses_allowed must be an integer or null", Codes.INVALID_PARAM
+            )
+
+        expiry_time = body.get("expiry_time", None)
+        if not isinstance(expiry_time, (int, type(None))):
+            raise SynapseError(
+                400, "expiry_time must be an integer or null", Codes.INVALID_PARAM
+            )
+
+        res = await self.store.create_registration_token(
+            token, uses_allowed, expiry_time
+        )
+        if not res:
+            # Creation failed, probably the token already exists
+            raise SynapseError(
+                400, "Token already exists: {}".format(token), Codes.INVALID_PARAM
+            )
+
+        resp = {
+            "token": token,
+            "uses_allowed": uses_allowed,
+            "pending": 0,
+            "completed": 0,
+            "expiry_time": expiry_time,
+        }
+        return 200, resp
+
+
+class RegistrationTokenRestServelet(RestServlet):
+    """Retrieve, update, or delete the given token."""
+
+    PATTERNS = admin_patterns("/registration_tokens/(?P<token>[^/]*)$")
+
+    def __init__(self, hs: "HomeServer"):
+        self.hs = hs
+        self.auth = hs.get_auth()
+        self.store = hs.get_datastore()
+
+    async def on_GET(self, request: SynapseRequest, token: str) -> Tuple[int, JsonDict]:
+        """Retrieve a registration token."""
+        await assert_requester_is_admin(self.auth, request)
+        token_info = await self.store.get_one_registration_token(token)
+
+        # If no result return a 404
+        if token_info is None:
+            raise NotFoundError("No such registration token: {}".format(token))
+
+        return 200, token_info
+
+    async def on_PUT(self, request: SynapseRequest, token: str) -> Tuple[int, JsonDict]:
+        """Update a registration token."""
+        await assert_requester_is_admin(self.auth, request)
+        body = parse_json_object_from_request(request)
+        updatevalues = {}
```

Comment:
Ah, this is just because we're passing to the `updatevalues` arg eventually.

Still, I'd suggest using something like `token_attributes` which may be a bit more descriptive.